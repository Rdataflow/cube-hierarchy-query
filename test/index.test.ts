import { meta } from '@zazuko/vocabulary-extras/builders'
import { schema, sh } from '@tpluscode/rdf-ns-builders'
import { expect } from 'chai'
import $rdf from 'rdf-ext'
import { getHierarchy, HierarchyNode } from '../index.js'
import { ex, parse, startFuseki } from './support.js'
import { insertGeoData } from './testData.js'
import { streamClient } from './client.js'

describe('@zazuko/cube-hierarchy-query', () => {
  before(startFuseki)

  before(insertGeoData)

  describe('getHierarchy', () => {
    const countriesHierarchy = parse`
      <>
        ${meta.hierarchyRoot} <Europe>, <North-America>, <South-America>, <Asia> ;
        ${meta.nextInHierarchy} <countryLevel> ;
      .
      
      <countryLevel>
        ${sh.path} [
          ${sh.inversePath} ${schema.containedInPlace} ;
        ] ;
        ${meta.nextInHierarchy} <cantonLevel> ;
      .
      
      <cantonLevel> 
        ${sh.path} [
          ${sh.inversePath} ${schema.containedInPlace} ;
        ] ;
        ${meta.nextInHierarchy} <municipalityLevel> ;
      .
      
      <municipalityLevel> 
        ${sh.path} [
          ${sh.inversePath} ${schema.containedInPlace} ;
        ] ;
        ${meta.nextInHierarchy} <districtLevel> ;
      .
      
      <districtLevel> ${sh.path} ${schema.containsPlace} .
    `

    it('loads all levels', async function () {
      // given
      const hierarchy = await countriesHierarchy

      // when
      const hierarchyQuery = getHierarchy(hierarchy.namedNode(ex('')))
      const hierarchyTree = await hierarchyQuery.execute(streamClient, $rdf)

      expect(hierarchyQuery.query.build()).to.matchSnapshot(this)

      const plainTree = hierarchyTree.map(toPlain)

      // then
      expect(plainTree).to.containSubset([
        {
          resource: ex('Europe'),
          name: 'Europe',
          nextInHierarchy: [
            {
              resource: ex('CH'),
              name: 'Switzerland',
              nextInHierarchy: [
                {
                  resource: ex('ZH'),
                  name: 'Kanton ZÃ¼rich',
                  nextInHierarchy: [
                    {
                      resource: ex('Affoltern'),
                      nextInHierarchy: [
                        { resource: ex('Bonstetten'), name: 'Bonstetten' },
                        { resource: ex('Rifferswil'), name: 'Rifferswil' },
                        { resource: ex('Stallikon'), name: 'Stallikon' },
                        { resource: ex('Knonau'), name: 'Knonau' },
                      ],
                    },
                  ],
                },
                {
                  resource: ex('BE'),
                  nextInHierarchy: [
                    {
                      resource: ex('Biel-Bienne'),
                      nextInHierarchy: [
                        { resource: ex('Aegerten') },
                        { resource: ex('Ligerz') },
                        { resource: ex('Nidau') },
                        { resource: ex('Port') },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
        {
          resource: ex('Asia'),
        },
      ])
    })

    it('loads shallower paths', async () => {
      // given
      const hierarchy = await countriesHierarchy

      // when
      const hierarchyTree = await getHierarchy(hierarchy.namedNode(ex(''))).execute(streamClient, $rdf)
      const plainTree = hierarchyTree.map(toPlain)

      // then
      expect(plainTree).to.containSubset([{
        resource: ex('North-America'),
        nextInHierarchy: [
          { resource: ex('US') },
        ],
      },
      {
        resource: ex('South-America'),
        nextInHierarchy: [
          { resource: ex('BR') },
          { resource: ex('AR') },
          { resource: ex('VE') },
        ],
      }])
    })

    it('applies type filter', async () => {
      // given
      const hierarchy = await countriesHierarchy
      hierarchy.namedNode(ex('countryLevel'))
        .addOut(sh.targetClass, ex.Country)

      // when
      const hierarchyTree = await getHierarchy(hierarchy.namedNode(ex(''))).execute(streamClient, $rdf)
      const plainTree = hierarchyTree.map(toPlain)

      // then
      expect(plainTree).not.to.containSubset([{
        resource: ex('Europe'),
        nextInHierarchy: [
          { resource: $rdf.namedNode('https://sws.geonames.org/798544') },
        ],
      }])
    })
  })
})

function toPlain(node: HierarchyNode): Record<string, unknown> {
  if (!node.nextInHierarchy.length) {
    return {
      resource: node.resource.term,
      name: node.resource.out(schema.name).value,
    }
  }

  return {
    resource: node.resource.term,
    nextInHierarchy: node.nextInHierarchy.map(toPlain),
    name: node.resource.out(schema.name).value,
  }
}
