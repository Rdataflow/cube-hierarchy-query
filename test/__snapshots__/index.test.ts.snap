exports["@zazuko/cube-hierarchy-query getHierarchy loads all levels"] = "PREFIX schema: <http://schema.org/>\n\nDESCRIBE <http://example.com/Europe>\n<http://example.com/North-America>\n<http://example.com/South-America>\n<http://example.com/Asia> ?level1\n?level2\n?level3\n?level4\n\nWHERE {\n\n    VALUES ( ?root )\n{\n( <http://example.com/Europe> )\n( <http://example.com/North-America> )\n( <http://example.com/South-America> )\n( <http://example.com/Asia> )\n}\n    \n    ?root ^schema:containedInPlace ?level1 .\n      OPTIONAL {\n        ?level1 ^schema:containedInPlace ?level2 .\n      OPTIONAL {\n        ?level2 ^schema:containedInPlace ?level3 .\n      OPTIONAL {\n        ?level3 schema:containsPlace ?level4 .\n      }\n    \n      }\n    \n      }\n    \n  \n}\n\n";

